content/shmem_init_thread.tex

```diff
@@ -0,0 +1,78 @@
+\apisummary{
+Initializes the \openshmem{} library, similar to SHMEM\_INIT(), and in addition
+perform the initialization required for thread-safe invocation of \openshmem{} functions.}
+
+\begin{apidefinition}
+
+\begin{Csynopsis}
+int shmem_init_thread(int requested, int *provided);
+\end{Csynopsis}
+
+\begin{Fsynopsis}  
+INTEGER requested, provided
+     
+SHMEM_INIT_THREAD(requested, provided)
+\end{Fsynopsis}  
+
+\begin{apiarguments}
+\apiargument{IN}{requested}  {The thread level support requested by the user.
+The correct values are SHMEM\_THREAD\_SINGLE, SHMEM\_THREAD\_FUNNELED, or SHMEM\_THREAD\_MULTIPLE}
+\apiargument{OUT}{provided}{The thread level support provided by the \openshmem{} implementation.}
+\end{apiarguments}
+
+
+\apidescription{
+    \FUNC{shmem\_init\_thread}  initializes the \openshmem{} library similar to
+    \FUNC{shmem\_init}, and in addition perform the initialization required for
+    thread-safe invocation of \openshmem{} functions. The argument
+    \textit{requested} is used to specify the desired level of thread support.
+    The function returns the support provided by the library. There are four 
+    levels of threading support.
+ 
+ \begin{itemize}
+\item SHMEM\_THREAD\_SINGLE: The \openshmem{} program may not be multithreaded, 
+and all \openshmem{} interfaces are invoked by a single thread. 
+
+\item SHMEM\_THREAD\_FUNNELED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces.
+
+\item SHMEM\_THREAD\_SERIALIZED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces at any given instance i.e., the \openshmem{} interfaces 
+are not invoked concurrently by multiple threads.
+
+\item SHMEM\_THREAD\_MULTIPLE: 
+The \openshmem{} program may be multithreaded, and any 
+thread may invoke the \openshmem{} interfaces.
+
+\item SHMEM\_THREAD\_UNSUPPORTED: The \openshmem{} specification 
+can be implemented without the support for threads, or a model beyond the four
+models.
+   
+\end{itemize}
+
+The function may be used to initialize \openshmem{}, and to initialize the
+\openshmem{} with thread safety, instead of SHMEM\_INIT. The SHMEM\_INIT\_THREAD
+may not be called multiple times in an \openshmem{} program.
```

Comment:
> Though the shmem_init(); shmem_init_thread() sequence usage model seems to provide some performance efficiency at present, in the Context time frame this would be too difficult to handle. The default property of all the contexts depends on the threading model value from shmem_init_thread(). Allowing contexts usage in @nspark example like;
> 
> shmem_init();
> shmem_ctx_create(0, ctx1); /* consider 0 as the default option */
> shmem_ctx_create(0, ctx2);
> shmem_init_thread();
> would require the implementation to change the property of all the contexts and rearrange the network mapping after shmem_init_thread(). This would be hard for the implementation to handle.
> @jdinan please clarify, if my understanding on the Context approach is correct.

@naveen-rn As long as, the context can be created in SHARED or PRIVATE mode, and can be used with/without threads, it should be enough to support this. I’m not sure, I understand the new challenge.